### PR TITLE
Compress alertmanager config

### LINKED ISSUE #4902: Hit Maximum secret size for alertmanager.yaml in kubernetes
**What happened?**
When creating an alertmanager config with recievers and routes for 300 "teams" within our organisation, we hit an error with the maximum secret size similar to the error with prometheus config in [this issue](https://github.com/prometheus-operator/prometheus-operator/issues/2234).

I see the following in the logs
```
level=error ts=2022-07-07T14:09:54.349591481Z caller=klog.go:116 component=k8s_client_runtime func=ErrorDepth msg="Sync \"monitoring/test\" failed: provision alertmanager configuration: create or update generated config secret failed: failed to update generated config secret: Secret \"alertmanager-test-generated\" is invalid: data: Too long: must have at most 1048576 bytes"
```

Would you be open to a similar fix to #2243 but for alertmanager.yaml if I was to raise a PR? i've done some tests and our generated config file is 1.4M, but gzipping it compresses to 28k - as the alertmanager configuration file we generate is very verbose

**Did you expect to see something different?**
I'd expect the config generation not to fail, or to reject the alertmanagerconfiguration if it is too large

**How to reproduce it (as minimally and precisely as possible)**:
Create a large alertmanagerconfig resource - ours is 1.6M - I may be able to provide a redacted version if this is useful

**Environment**

* Prometheus Operator version:

    `v0.53.1`
    
    

* Kubernetes version information:

    `Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:31:32Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.10-eks-84b4fe6", GitCommit:"cc6a1b4915a99f49f5510ef0667f94b9ca832a8a", GitTreeState:"clean", BuildDate:"2022-06-09T18:24:04Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}`
    

* Kubernetes cluster kind:

EKS

* Manifests:
I can provide a redacted version of our alertmanagerconfig resource if this is useful
```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
level=error ts=2022-07-07T14:09:54.349591481Z caller=klog.go:116 component=k8s_client_runtime func=ErrorDepth msg="Sync \"monitoring/test\" failed: provision alertmanager configuration: create or update generated config secret failed: failed to update generated config secret: Secret \"alertmanager-test-generated\" is invalid: data: Too long: must have at most 1048576 bytes"
```

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by timmow**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Fixes #4902 - more details in the issue

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
compress alertmanager config secret data to avoid hitting maximum secret size with large alertmanagerconfig objects
```

### CODE REVIEW

File: pkg/alertmanager/statefulset.go [hunk 1]
```diff
@@ -488,13 +502,17 @@ func makeStatefulSetSpec(a *monitoringv1.Alertmanager, config Config, tlsAssetSe
 		},
 	}
 
-	reloadWatchDirs := []string{alertmanagerConfigDir}
+	reloadWatchDirs := []string{}
```
Comment (simonpasquier) [inline pkg/alertmanager/statefulset.go:505]: I'm not sure why we had `reloadWatchDirs` and `watchedDirectories` while they are identical copies... Let's get rid of `reloadWatchDirs`.
 
```suggestion
	var watchedDirectories []string
```

**Review by simonpasquier — COMMENTED** [review]
Cool! Just one small comment to avoid unnecessary variables otheriwse I think that you can switch off Draft mode :)

**Comment by timmow** [comment]
i've left individual commits for ease of reviewing - do you want me to squash these into 1 commit / rebase against `main` ?

**Comment by PhilipGough** [comment]
> i've left individual commits for ease of reviewing - do you want me to squash these into 1 commit / rebase against main ?

We can squash on merge so happy to review as is. Thanks

**Review by simonpasquier — APPROVED** [review]

**Review by PhilipGough — APPROVED** [review]
lgtm

**Pull request merged**